vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/samples/registration/batch/SampleRegistrationSpreadsheet.java

```diff
@@ -398,42 +405,78 @@ public List<SamplesheetHeaderName> retrieveGenomics() {
+      String speciesInput = parseAndTrimCellOfRow(row, header.indexOf(SamplesheetHeaderName.SPECIES));
+      String specimenInput = parseAndTrimCellOfRow(row, header.indexOf(SamplesheetHeaderName.SPECIMEN));
+      String analyteInput = parseAndTrimCellOfRow(row, header.indexOf(SamplesheetHeaderName.ANALYTE));
+      String commentInput = parseAndTrimCellOfRow(row, header.indexOf(SamplesheetHeaderName.CUSTOMER_COMMENT));
+
+      // we need to stream this list twice, so we use a supplier
+      Supplier<Stream<String>> mandatoryCellStreamSupplier = () -> Stream.of(analysisTypeInput,
+          sampleLabelInput,
+          replicateIDInput, conditionInput, speciesInput, specimenInput, analyteInput);
 
       if (mandatoryCellStreamSupplier.get().anyMatch(Objects::isNull)) {
         break;
       }
-      if (mandatoryCellStreamSupplier.get().noneMatch(x -> x.getStringCellValue().isEmpty())) {
-        ExperimentalGroup experimentalGroup = experimentalGroupToConditionString.get(
-            conditionCell.getStringCellValue().trim());
-        Long experimentalGroupId = experimentalGroup.id();
-        BiologicalReplicateId biologicalReplicateId = retrieveBiologicalReplicateId(
-            replicateIDCell.getStringCellValue().trim(), conditionCell.getStringCellValue().trim());
-        rows.add(new NGSRowDTO(analysisTypeCell.getStringCellValue().trim(),
-            sampleLabelCell.getStringCellValue().trim(),
-            biologicalReplicateId, experimentalGroupId,
-            speciesCell.getStringCellValue().trim(), specimenCell.getStringCellValue().trim(),
-            analyteCell.getStringCellValue().trim(),
-            commentCell.getStringCellValue().trim()));
+
+      if(mandatoryCellStreamSupplier.get().noneMatch(x -> x.isEmpty())) {
+        String uniqueSampleString = replicateIDInput+conditionInput;
+        if(uniqueSamples.contains(uniqueSampleString)) {
+          return Result.fromError(new SpreadsheetValidationException(
+              "Biological replicate Id "+replicateIDInput+" was used multiple times for the same condition.", rowId));
+        } else {
+          ExperimentalGroup experimentalGroup = experimentalGroupToConditionString.get(
+              conditionInput);
+          Long experimentalGroupId = experimentalGroup.id();
+          BiologicalReplicateId biologicalReplicateId = retrieveBiologicalReplicateId(
+              replicateIDInput, conditionInput);
+          rows.add(new NGSRowDTO(analysisTypeInput, sampleLabelInput,
+              biologicalReplicateId, experimentalGroupId,
+              speciesInput, specimenInput, analyteInput,
+              commentInput));
+          uniqueSamples.add(uniqueSampleString);
+        }
+      } else {
+        return Result.fromError(new SpreadsheetValidationException(
+            "Missing mandatory fields.", rowId));
+      }
+
+    }
+    return Result.fromValue(rows);
+  }
```

Comment:
This method is extremely long and quite complex. Can you break it down? It is a lot to stomach.

Reply:
It is now broken down into validation and another method returning the actual rows. they are still long, but most of this is due to having to access multiple columns.

Reply:
looking at your comments, I see some possible method extractions.
`break when cells are undefined`
`mandatory not filled in -> invalid`
`sample uniqueness needs to be guaranteed by condition and replicate ID`